rfd/0006-reverse-tunneling.md

```diff
@@ -0,0 +1,39 @@
+---
+authors: Adam Enger (adamenger@gmail.com)
+state: discussion
+---
+
+# RFD 6 - Reverse Tunneling
+
+## What
+
+Implementation of [RFC4254 Section 7.1, Reverse Port Forwarding](https://tools.ietf.org/html/rfc4254#section-7.1)
+
+## Why
+
+Teleport supports Local Port Forwarding but does not support Reverse Port Forwarding. SSH supports the ability to route remote connections to a local listener. Implementing Reverse Port Forwarding will expand the possible use cases for `tsh` and bridge the gap between `ssh` and `tsh` functionality.
+
+## Details
+
+Unlike Local Tunnels, a few things need to be modified on the server side in order to handle incoming ssh forwarding requests.
+
+### Global Requests
+
+Reverse Tunneling requires that the SSH server handles requests of type `tcpip-forward` and `cancel-tcpip-forward`. We will need to expand the global request handler to handle both of these request types and take the appropriate actions when they are encountered.
+
+#### tcpip-forward requests
+
+These requests are sent by the client when the client asks the remote end to open a port. The server should validate whether or not the user has permissions to port forward first. If the user has permission, a listening socket will be opened on the Node and an SSH channel of type `forwarded-tcpip` from the Node to the Client is opened. 
+
+#### cancel-tcpip-forward requests
+
+These requests are sent when the client requests port forwarding to be canceled. The server should respond to these messages with TRUE which will close the remote channel. *NOTE:* channel open requests may be received until a reply to this message is received.
+
+### tsh client
+
+Minor modifications are required to the client to support Reverse Tunneling. We will need to add an additional command line argument and an additional case for Remote Port Forwarding in the startPortForwarding function.
+
```

Comment:
Any changes to our RBAC? Will this follow `port-forward` RBAC option?

Reply:
I think this should be bundled under the `allow_port_forwarding` RBAC rule we have already. Splitting them into `allow_local_port_forwarding` and `allow_remote_port_forwarding` is probably superfluous, given that any port forward is somewhat bidirectional anyway.

Reply:
Do you have an example in the codebase for how this rule is checked? I can update the RFD and client code for reverse tunneling to respect `allow_port_forwarding`.

Reply:
This is the function in the auth code that checks whether it's allowed or not: https://github.com/gravitational/teleport/blob/e64f399cffcbd31d0c7e3c5f88db35480836930f/lib/srv/authhandlers.go#L105-L140

This function is called in the regular Teleport SSH server here: https://github.com/gravitational/teleport/blob/e64f399cffcbd31d0c7e3c5f88db35480836930f/lib/srv/regular/sshserver.go#L1066-L1071

It's also called for the forwarding SSH server (which is used for proxy recording mode) here: https://github.com/gravitational/teleport/blob/e64f399cffcbd31d0c7e3c5f88db35480836930f/lib/srv/forward/sshserver.go#L683-L688